common/persistence/cassandra/cassandraQueue.go

```diff
@@ -0,0 +1,200 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cassandra
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gocql/gocql"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/service/config"
+)
+
+const (
+	templateEnqueueMessageQuery   = `INSERT INTO queue (queue_type, message_id, message_payload) VALUES(?, ?, ?) IF NOT EXISTS`
+	templateGetLastMessageIDQuery = `SELECT message_id FROM queue WHERE queue_type=? ORDER BY message_id DESC LIMIT 1`
+	templateGetMessagesQuery      = `SELECT message_id, message_payload ` +
+		`FROM queue ` +
+		`WHERE queue_type = ? ` +
+		`and message_id > ? `
+)
+
+type (
+	cassandraQueue struct {
+		queueType   int
+		logger      log.Logger
+		retryPolicy backoff.RetryPolicy
+		cassandraStore
+	}
+
+	messageIDConflictError struct {
+		messageID int
+	}
+)
+
+func newQueue(
+	cfg config.Cassandra,
+	logger log.Logger,
+	queueType int,
+) (persistence.Queue, error) {
+	cluster := NewCassandraCluster(cfg.Hosts, cfg.Port, cfg.User, cfg.Password, cfg.Datacenter)
+	cluster.Keyspace = cfg.Keyspace
+	cluster.ProtoVersion = cassandraProtoVersion
+	cluster.Consistency = gocql.LocalQuorum
+	cluster.SerialConsistency = gocql.LocalSerial
+	cluster.Timeout = defaultSessionTimeout
+
+	session, err := cluster.CreateSession()
+	if err != nil {
+		return nil, err
+	}
+
+	retryPolicy := backoff.NewExponentialRetryPolicy(100 * time.Millisecond)
+	retryPolicy.SetBackoffCoefficient(1.5)
+	retryPolicy.SetMaximumAttempts(5)
+
+	return &cassandraQueue{
+		cassandraStore: cassandraStore{session: session, logger: logger},
+		logger:         logger,
+		queueType:      queueType,
+		retryPolicy:    retryPolicy,
+	}, nil
+}
+
+func (q *cassandraQueue) EnqueueMessage(messagePayload []byte) error {
+	err := backoff.Retry(
```

Comment:
I don't think we need backoff retry policy here. Persistence API are expected to be very low latency, so that it's much easier for us to troubleshoot with Cassandra. the retry should be in application layer. So the number requests sent to Cassandra should be constant and deterministic.

Reply:
ID conflict error is implementation detail that should be hide inside queue implementation. Caller should not try to parse and understand this error.

Reply:
I agree with @longquanzheng persistence layer should only handle converting to / from DB
no business logic, or retry logic should be put here